components/raftstore-v2/src/operation/life.rs

```diff
@@ -260,7 +306,225 @@ impl Store {
     }
 }
 
+/// Tell leader that `to_peer` from `tombstone_msg` is destroyed.
+fn build_peer_destroyed_report(tombstone_msg: &mut RaftMessage) -> Option<RaftMessage> {
+    let to_region_id = if tombstone_msg.has_extra_msg() {
+        assert_eq!(
+            tombstone_msg.get_extra_msg().get_type(),
+            ExtraMessageType::MsgGcPeerRequest
+        );
+        tombstone_msg
+            .get_extra_msg()
+            .get_check_gc_peer()
+            .get_from_region_id()
+    } else {
+        tombstone_msg.get_region_id()
+    };
+    if to_region_id == 0 || tombstone_msg.get_from_peer().get_id() == 0 {
+        return None;
+    }
+    let mut msg = RaftMessage::default();
+    msg.set_region_id(to_region_id);
+    msg.set_from_peer(tombstone_msg.take_to_peer());
+    msg.set_to_peer(tombstone_msg.take_from_peer());
+    msg.mut_extra_msg()
+        .set_type(ExtraMessageType::MsgGcPeerResponse);
+    Some(msg)
+}
+
+/// Forward the destroy request from target peer to merged source peer.
+fn forward_destroy_source_peer<EK, ER, T>(ctx: &mut StoreContext<EK, ER, T>, msg: &RaftMessage)
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+    T: Transport,
+{
+    let extra_msg = msg.get_extra_msg();
+    // Instead of respond leader directly, send a message to target region to
+    // double check it's really destroyed.
+    let check_gc_peer = extra_msg.get_check_gc_peer();
+    let mut tombstone_msg = Box::<RaftMessage>::default();
+    tombstone_msg.set_region_id(check_gc_peer.get_check_region_id());
+    tombstone_msg.set_from_peer(msg.get_from_peer().clone());
+    tombstone_msg.set_to_peer(check_gc_peer.get_check_peer().clone());
+    tombstone_msg.set_region_epoch(check_gc_peer.get_check_region_epoch().clone());
+    tombstone_msg.set_is_tombstone(true);
+    // No need to set epoch as we don't know what it is.
+    tombstone_msg
+        .mut_extra_msg()
+        .set_type(ExtraMessageType::MsgGcPeerRequest);
```

Comment:
I think this should be `MsgGcPeerResponse`.

Reply:
No, it should be `MsgGcPeerRequest`. As the function name and comment says, the request is forwarded to source peer.

Reply:
Forget it, I misread.

Reply:
What is confusing though, is the use of `MsgGcPeerRequest`, when in fact only the `set_is_tombstone` is needed. For a reader reading `MsgGcPeerRequest`, he naturally wants to know the handling of it by skipping to `on_gc_peer_request`, but that function only makes sense for target peer not source peer.

Reply:
`MsgGcPeerRequest` is necessary. `is_tombstone` is used by regular destroy, which indicates `from` is in the same region as `to`. `MsgGcPeerRequest` is for merge destroy, `from` may not be in the same region as `to`.